include/nbl/builtin/hlsl/spirv_intrinsics/core.hlsl

```diff
@@ -19,26 +18,49 @@ namespace hlsl
 #ifdef __HLSL_VERSION
 namespace spirv
 {
-//! General
-[[vk::ext_builtin_input(spv::BuiltInHelperInvocation)]]
-static const bool HelperInvocation;
 
-//! Vertex Inputs
-[[vk::ext_builtin_input(spv::BuiltInVertexIndex)]]
-static const uint32_t VertexIndex;
-[[vk::ext_builtin_input(spv::BuiltInInstanceIndex)]]
-static const uint32_t InstanceIndex;
+//! General Decls
+template<uint32_t StorageClass, typename T>
+using pointer_t = vk::SpirvOpaqueType<spv::OpTypePointer, vk::Literal< vk::integral_constant<uint32_t, StorageClass> >, T>;
 
-//! Vertex and friends
-[[vk::ext_builtin_output(spv::BuiltInPosition)]]
-static float32_t4 Position;
+// The holy operation that makes addrof possible
+template<uint32_t StorageClass, typename T>
+[[vk::ext_instruction(spv::OpCopyObject)]]
+pointer_t<StorageClass, T> copyObject([[vk::ext_reference]] T value);
+
+//! Std 450 Extended set operations
+template<typename SquareMatrix>
+[[vk::ext_instruction(GLSLstd450MatrixInverse)]]
+SquareMatrix matrixInverse(NBL_CONST_REF_ARG(SquareMatrix) mat);
+
+// Add specializations if you need to emit a `ext_capability` (this means that the instruction needs to forward through an `impl::` struct and so on)
+template<typename T, typename U>
+[[vk::ext_capability(spv::CapabilityPhysicalStorageBufferAddresses)]]
+[[vk::ext_instruction(spv::OpBitcast)]]
+enable_if_t<is_spirv_type_v<T> && is_spirv_type_v<U>, T> bitcast(U);
+
+template<typename T>
+[[vk::ext_capability(spv::CapabilityPhysicalStorageBufferAddresses)]]
+[[vk::ext_instruction(spv::OpBitcast)]]
+uint64_t bitcast(pointer_t<spv::StorageClassPhysicalStorageBuffer,T>);
+
+template<typename T>
+[[vk::ext_capability(spv::CapabilityPhysicalStorageBufferAddresses)]]
+[[vk::ext_instruction(spv::OpBitcast)]]
+pointer_t<spv::StorageClassPhysicalStorageBuffer,T> bitcast(uint64_t);
 
-//! Compute Shader Builtins
+template<class T, class U>
+[[vk::ext_instruction(spv::OpBitcast)]]
+T bitcast(U);
```

Comment:
needs enable_if to check that `T` is fundamental or a builtin HLSL `vector` (read the SPIR-V spec) + `sizeof(T)==sizeof(U)`